// @nx/react/src/module-federation/with-module-federation-ssr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withModuleFederationForSSR = void 0;
const utils_1 = require("./utils");
async function withModuleFederationForSSR(options) {
    const { sharedLibraries, sharedDependencies, mappedRemotes } = await (0, utils_1.getModuleFederationConfig)(options, {
        isServer: true,
    });
    return (config) => {
        config.target = false;
        config.output.uniqueName = options.name;
        config.optimization = {
            runtimeChunk: false,
        };
        config.plugins.push(new (require('@module-federation/node').UniversalFederationPlugin)({
            name: options.name,
            filename: 'remoteEntry.js',
            exposes: options.exposes,
            remotes: mappedRemotes,
            shared: {
                ...sharedDependencies,
            },
            library: {
                type: 'commonjs-module',
            },
            isServer: true,
        }, {}), sharedLibraries.getReplacementPlugin());
        return config;
    };
}
exports.withModuleFederationForSSR = withModuleFederationForSSR;
