// @nx/next/src/generators/application/lib/add-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addProject = void 0;
const devkit_1 = require("@nx/devkit");
const add_build_target_defaults_1 = require("@nx/devkit/src/generators/add-build-target-defaults");
function addProject(host, options) {
    const targets = {};
    // Check if plugin exists in nx.json and if it doesn't then we can continue
    // with the default targets.
    const nxJson = (0, devkit_1.readNxJson)(host);
    const hasPlugin = nxJson.plugins?.some((p) => typeof p === 'string'
        ? p === '@nx/next/plugin'
        : p.plugin === '@nx/next/plugin');
    if (!hasPlugin) {
        (0, add_build_target_defaults_1.addBuildTargetDefaults)(host, '@nx/next:build');
        targets.build = {
            executor: '@nx/next:build',
            outputs: ['{options.outputPath}'],
            defaultConfiguration: 'production',
            options: {
                outputPath: options.outputPath,
            },
            configurations: {
                development: {
                    outputPath: options.appProjectRoot,
                },
                production: {},
            },
        };
        targets.serve = {
            executor: '@nx/next:server',
            defaultConfiguration: 'development',
            options: {
                buildTarget: `${options.projectName}:build`,
                dev: true,
            },
            configurations: {
                development: {
                    buildTarget: `${options.projectName}:build:development`,
                    dev: true,
                },
                production: {
                    buildTarget: `${options.projectName}:build:production`,
                    dev: false,
                },
            },
        };
        targets.export = {
            executor: '@nx/next:export',
            options: {
                buildTarget: `${options.projectName}:build:production`,
            },
        };
    }
    const project = {
        root: options.appProjectRoot,
        sourceRoot: options.appProjectRoot,
        projectType: 'application',
        targets,
        tags: options.parsedTags,
    };
    (0, devkit_1.addProjectConfiguration)(host, options.projectName, {
        ...project,
    });
}
exports.addProject = addProject;
